api/eth_types.go

```diff
@@ -0,0 +1,442 @@
+
+var (
+	EthBigIntZero = EthBigInt{Int: big.Zero().Int}
+)
+
+func (e EthBigInt) MarshalJSON() ([]byte, error) {
+	if e.Int == nil {
+		return json.Marshal("0x0")
+	}
+	return json.Marshal(fmt.Sprintf("0x%x", e.Int))
+}
+
+func (e *EthBigInt) UnmarshalJSON(b []byte) error {
+	var s string
+	if err := json.Unmarshal(b, &s); err != nil {
+		return err
+	}
+
+	replaced := strings.Replace(s, "0x", "", -1)
+	if len(replaced)%2 == 1 {
+		replaced = "0" + replaced
+	}
+
+	i := new(mathbig.Int)
+	i.SetString(replaced, 16)
+
+	*e = EthBigInt(big.NewFromGo(i))
+	return nil
+}
+
+type EthBytes []byte
+
+func (e EthBytes) MarshalJSON() ([]byte, error) {
+	if len(e) == 0 {
+		return json.Marshal("0x00")
+	}
+	s := hex.EncodeToString(e)
+	if len(s)%2 == 1 {
+		s = "0" + s
+	}
+	return json.Marshal("0x" + s)
+}
+
+func (e *EthBytes) UnmarshalJSON(b []byte) error {
+	var s string
+	if err := json.Unmarshal(b, &s); err != nil {
+		return err
+	}
+
+	s = strings.Replace(s, "0x", "", -1)
+	if len(s)%2 == 1 {
+		s = "0" + s
+	}
+
+	decoded, err := hex.DecodeString(s)
+	if err != nil {
+		return err
+	}
+
+	*e = decoded
+	return nil
+}
+
+type EthBlock struct {
+	ParentHash       EthHash    `json:"parentHash"`
+	Sha3Uncles       EthHash    `json:"sha3Uncles"`
+	Miner            EthAddress `json:"miner"`
+	StateRoot        EthHash    `json:"stateRoot"`
+	TransactionsRoot EthHash    `json:"transactionsRoot"`
+	ReceiptsRoot     EthHash    `json:"receiptsRoot"`
+	// TODO: include LogsBloom
+	Difficulty    EthUint64 `json:"difficulty"`
+	Number        EthUint64 `json:"number"`
+	GasLimit      EthUint64 `json:"gasLimit"`
+	GasUsed       EthUint64 `json:"gasUsed"`
+	Timestamp     EthUint64 `json:"timestamp"`
+	Extradata     []byte    `json:"extraData"`
+	MixHash       EthHash   `json:"mixHash"`
+	Nonce         EthNonce  `json:"nonce"`
+	BaseFeePerGas EthBigInt `json:"baseFeePerGas"`
+	Size          EthUint64 `json:"size"`
+	// can be []EthTx or []string depending on query params
+	Transactions []interface{} `json:"transactions"`
+	Uncles       []EthHash     `json:"uncles"`
+}
+
+var (
+	EmptyEthHash  = EthHash{}
+	EmptyEthInt   = EthUint64(0)
+	EmptyEthNonce = [8]byte{0, 0, 0, 0, 0, 0, 0, 0}
+)
+
+func NewEthBlock() EthBlock {
+	return EthBlock{
+		Sha3Uncles:       EmptyEthHash,
+		StateRoot:        EmptyEthHash,
+		TransactionsRoot: EmptyEthHash,
+		ReceiptsRoot:     EmptyEthHash,
+		Difficulty:       EmptyEthInt,
+		Extradata:        []byte{},
+		MixHash:          EmptyEthHash,
+		Nonce:            EmptyEthNonce,
+		GasLimit:         EthUint64(build.BlockGasLimit), // TODO we map Ethereum blocks to Filecoin tipsets; this is inconsistent.
+		Uncles:           []EthHash{},
+		Transactions:     []interface{}{},
+	}
+}
+
+type EthCall struct {
+	From     *EthAddress `json:"from"`
+	To       *EthAddress `json:"to"`
+	Gas      EthUint64   `json:"gas"`
+	GasPrice EthBigInt   `json:"gasPrice"`
+	Value    EthBigInt   `json:"value"`
+	Data     EthBytes    `json:"data"`
+}
```

Comment:
- nonce
- condition?

https://openethereum.github.io/JSONRPC.html#transactions

Reply:
Nope, those are OpenEthereum-specific. See https://docs.infura.io/infura/networks/ethereum/json-rpc-methods/eth_call

However, we may need to allow these optional fields:
- maxPriorityFeePerGas
- maxFeePerGas

Not sure why they are accepted as part of eth_call or eth_estimateGas.

Reply:
I don't think they are needed here, because they'll be discarded anyway when we turn this EthCall into a Filecoin message. It won't cause any compatibility issues because users can specify those additional parameters if they want, it's just that they won't be used.